src/lib/shadow-shim-helper-rs/src/syscall_types.rs

```diff
@@ -238,15 +322,17 @@ impl TryFrom<SysCallReg> for i16 {
     }
 }
 
-impl From<ForeignPtr> for SysCallReg {
-    fn from(v: ForeignPtr) -> Self {
-        Self { as_ptr: v }
+impl<T> From<ForeignPtr<T>> for SysCallReg {
+    fn from(v: ForeignPtr<T>) -> Self {
+        Self {
+            as_ptr: v.cast::<()>(),
+        }
     }
 }
 
-impl From<SysCallReg> for ForeignPtr {
-    fn from(v: SysCallReg) -> ForeignPtr {
-        unsafe { v.as_ptr }
+impl<T> From<SysCallReg> for ForeignPtr<T> {
```

Comment:
It does infer the type, which normally I wanted to avoid, but we need it so that `SyscallHandlerFn` can convert the `SysCallReg` to the correct pointer type. For example in the following code, if a syscall handler has an argument of type `ForeignPtr<u32>`, the `SyscallHandlerFn` trait needs to be able to convert from the syscall register to the `ForeignPtr<u32>`.

https://github.com/shadow/shadow/blob/5efe48f66d107d33f86fe95d50950faea504f524/src/main/host/syscall/handler/mod.rs#L171-L181

I'm not sure of a better way we can handle this.

Reply:
Ah I see - yeah I don't see a better approach. A comment here for future reference would be helpful.

Reply:
Added a comment.